File: scripts/update_rules.py

import os
import re
import shutil
import sys

print("Starting script...")
MAX_CHARS = 12000

mapping = {
    "ab-testing-docs.md": ["docs/architecture/ab-testing.md"],
    "admin-api-docs.md": ["docs/architecture/admin-api.md"],
    "admin-controller-docs.md": ["docs/modules/server/admin-controller.md"],
    "agents-md.md": ["AGENTS.md"],
    "architecture-docs.md": ["docs/ARCHITECTURE.md"],
    "codebase-map.md": ["docs/CODEBASE_MAP.md"],
    "component-cache-docs.md": ["docs/modules/server/component-cache.md"],
    "credential-storage-docs.md": ["docs/modules/utils/credential-store.md"],
    "credential-storage-guide.md": ["docs/CREDENTIAL_STORAGE.md"],
    "llm-client-docs.md": ["docs/modules/llm/client.md"],
    "llm-factory-docs.md": ["docs/modules/llm/factory.md"],
    "llm-pipeline-docs.md": ["docs/architecture/llm-pipeline.md"],
    "llm-provider-docs.md": [os.path.join("docs/modules/llm/providers", f) for f in os.listdir("docs/modules/llm/providers") if f.endswith(".md")],
    "macos-app-docs.md": ["docs/macos-app.md"],
    "notarization-docs.md": ["docs/NOTARIZATION.md"],
    "rest-api-controller-docs.md": ["docs/modules/server/rest-api-controller.md"],
    "runtime-config-docs.md": ["docs/modules/config/runtime-config.md"],
    "server-core-docs.md": ["docs/modules/server/server.md"],
    "session-lifecycle-docs.md": ["docs/architecture/session-lifecycle.md"],
    "session-store-docs.md": ["docs/modules/server/session-store.md"],
    "token-optimization-docs.md": ["docs/architecture/token-optimization.md"],
    "transition-ux-docs.md": ["docs/architecture/transition-ux.md"],
    "virtual-rest-api-docs.md": ["docs/architecture/virtual-rest-api.md"]
}

base_dir = "/Users/torben/VSCode/serve-llm"
rules_dir = os.path.join(base_dir, ".agent/rules")

COMMON_FRONTMATTER = """---
trigger: always_on
globs: **/*
---
"""

def split_content(text, limit):
    chunks = []
    while len(text) > limit:
        split_idx = text.rfind("\n", 0, limit)
        if split_idx == -1:
            split_idx = limit
        chunks.append(text[:split_idx])
        text = text[split_idx:]
    chunks.append(text)
    return chunks

print(f"Processing {len(mapping)} rules...")

for rule_file, doc_files in mapping.items():
    rule_path = os.path.join(rules_dir, rule_file)
    print(f"Processing {rule_file}...")
    
    frontmatter = COMMON_FRONTMATTER

    full_doc_content = ""
    for doc_file in doc_files:
        doc_path = os.path.join(base_dir, doc_file)
        if os.path.exists(doc_path):
            with open(doc_path, "r") as f:
                doc_text = f.read()
            if "ARCHITECTURE.md" in doc_file or rule_file == "read-architecture-docs.md":
                doc_text = doc_text.replace("](docs/", "](../../docs/")
            full_doc_content += f"\n# Content from {doc_file}\n\n{doc_text}\n"
        else:
            print(f"Warning: Doc file {doc_file} not found")

    chunk_size = MAX_CHARS - len(frontmatter) - 100
    chunks = split_content(full_doc_content, chunk_size)

    base_name = os.path.splitext(rule_file)[0]
    ext = os.path.splitext(rule_file)[1]

    for i, chunk in enumerate(chunks):
        if len(chunks) > 1:
            filename = f"{base_name}-part{i+1}{ext}"
        else:
            filename = rule_file
        
        out_path = os.path.join(rules_dir, filename)
        with open(out_path, "w") as f:
            f.write(frontmatter + "\n" + chunk)
        print(f"Written {filename} ({len(chunk)} chars)")

print("Copying AGENTS.md and ARCHITECTURE.md...")
shutil.copy(os.path.join(base_dir, "AGENTS.md"), os.path.join(rules_dir, "AGENTS.md"))
shutil.copy(os.path.join(base_dir, "docs/ARCHITECTURE.md"), os.path.join(rules_dir, "ARCHITECTURE.md"))

arch_rule_path = os.path.join(rules_dir, "ARCHITECTURE.md")
with open(arch_rule_path, "r") as f:
    arch_content = f.read()
arch_content = arch_content.replace("](docs/", "](../../docs/")
with open(arch_rule_path, "w") as f:
    f.write(arch_content)

print("Done.")

File: scripts/update-rules.ts
/*
type: new file
fileName: scripts/update-rules.ts
*/
import { readFileSync, writeFileSync, existsSync, readdirSync } from "node:fs";
import { join, dirname, extname, basename } from "node:path";
import { fileURLToPath } from "node:url";

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);
const PROJECT_ROOT = join(__dirname, "..");
const RULES_DIR = join(PROJECT_ROOT, ".agent/rules");
const MAX_CHARS = 12000;

interface RuleConfig {
    docs: string[];
    trigger: "always_on" | "glob";
    globs?: string[];
}

const CONFIGURATION: Record<string, RuleConfig> = {
    "CODEBASE_MAP.md": {
        docs: ["docs/CODEBASE_MAP.md"],
        trigger: "always_on",
    },
    "agents-md.md": {
        docs: ["AGENTS.md"],
        trigger: "always_on",
    },
    "architecture-docs.md": {
        docs: ["docs/ARCHITECTURE.md"],
        trigger: "always_on",
    },
    "ab-testing-docs.md": {
        docs: ["docs/architecture/ab-testing.md"],
        trigger: "glob",
        globs: [
            "src/server/session-store.ts",
            "src/server/admin-controller.ts",
            "frontend/src/components/ABWorkspaceShell.tsx",
            "frontend/src/pages/AbTestWorkspacePage.tsx",
            "frontend/src/interceptor*.ts",
        ],
    },
    "admin-api-docs.md": {
        docs: ["docs/architecture/admin-api.md"],
        trigger: "glob",
        globs: [
            "src/server/admin-controller.ts",
            "src/types/admin-api.ts",
            "frontend/src/api/admin.ts",
            "frontend/src/pages/AdminDashboard.tsx",
        ],
    },
    "admin-controller-docs.md": {
        docs: ["docs/modules/server/admin-controller.md"],
        trigger: "glob",
        globs: ["src/server/admin-controller.ts"],
    },
    "codebase-map.md": {
        docs: ["docs/CODEBASE_MAP.md"],
        trigger: "glob",
        globs: ["package.json", "src/**", "frontend/src/**"],
    },
    "component-cache-docs.md": {
        docs: ["docs/modules/server/component-cache.md"],
        trigger: "glob",
        globs: ["src/server/component-cache.ts", "src/server/server.ts"],
    },
    "credential-storage-docs.md": {
        docs: ["docs/modules/utils/credential-store.md"],
        trigger: "glob",
        globs: ["src/utils/credential-store.ts"],
    },
    "credential-storage-guide.md": {
        docs: ["docs/CREDENTIAL_STORAGE.md"],
        trigger: "glob",
        globs: [
            "src/utils/credential-store.ts",
            "src/server/server.ts",
            "src/config/runtime-config.ts",
        ],
    },
    "llm-client-docs.md": {
        docs: ["docs/modules/llm/client.md"],
        trigger: "glob",
        globs: ["src/llm/client.ts", "src/llm/*-client.ts"],
    },
    "llm-factory-docs.md": {
        docs: ["docs/modules/llm/factory.md"],
        trigger: "glob",
        globs: ["src/llm/factory.ts"],
    },
    "llm-pipeline-docs.md": {
        docs: ["docs/architecture/llm-pipeline.md"],
        trigger: "glob",
        globs: [
            "src/server/server.ts",
            "src/llm/client.ts",
            "src/views/loading-shell.ts",
            "src/views/loading-shell/**/*",
        ],
    },
    "llm-provider-docs.md": {
        docs: [], // Will be populated dynamically from docs/modules/llm/providers/
        trigger: "glob",
        globs: ["src/llm/*-client.ts", "src/llm/providers/*.ts"],
    },
    "macos-app-docs.md": {
        docs: ["docs/macos-app.md"],
        trigger: "glob",
        globs: ["scripts/macos-app/**", "scripts/*.sh", "scripts/entitlements.plist"],
    },
    "notarization-docs.md": {
        docs: ["docs/NOTARIZATION.md"],
        trigger: "glob",
        globs: [
            "scripts/notarize-macos-sea.sh",
            "scripts/verify-notarization.sh",
            "scripts/check-notarization-credentials.sh",
        ],
    },
    "rest-api-controller-docs.md": {
        docs: ["docs/modules/server/rest-api-controller.md"],
        trigger: "glob",
        globs: ["src/server/rest-api-controller.ts"],
    },
    "runtime-config-docs.md": {
        docs: ["docs/modules/config/runtime-config.md"],
        trigger: "glob",
        globs: [
            "src/config/runtime-config.ts",
            "src/cli/args.ts",
            "src/utils/config-store.ts",
        ],
    },
    "server-core-docs.md": {
        docs: ["docs/modules/server/server.md"],
        trigger: "glob",
        globs: ["src/server/server.ts", "src/index.ts"],
    },
    "session-lifecycle-docs.md": {
        docs: ["docs/architecture/session-lifecycle.md"],
        trigger: "glob",
        globs: ["src/server/session-store.ts", "src/types.ts"],
    },
    "session-store-docs.md": {
        docs: ["docs/modules/server/session-store.md"],
        trigger: "glob",
        globs: ["src/server/session-store.ts"],
    },
    "token-optimization-docs.md": {
        docs: ["docs/architecture/token-optimization.md"],
        trigger: "glob",
        globs: ["src/server/component-cache.ts", "src/server/server.ts"],
    },
    "transition-ux-docs.md": {
        docs: ["docs/architecture/transition-ux.md"],
        trigger: "glob",
        globs: [
            "src/views/loading-shell/**",
            "frontend/public/vaporvibe-interceptor-sw.js",
            "frontend/src/interceptor.ts",
        ],
    },
    "virtual-rest-api-docs.md": {
        docs: ["docs/architecture/virtual-rest-api.md"],
        trigger: "glob",
        globs: ["src/server/rest-api-controller.ts", "src/types.ts"],
    },
    "llm-test-architecture.md": {
        docs: ["tests/llm/README.md"],
        trigger: "glob",
        globs: [
            "tests/llm/**/*.test.ts",
            "tests/fixtures/**/*.json",
            "scripts/fixtures/**/*.ts",
        ],
    },
};

// Helper to dynamically populate provider docs
const providersDir = join(PROJECT_ROOT, "docs/modules/llm/providers");
if (existsSync(providersDir)) {
    const providerDocs = readdirSync(providersDir)
        .filter((f) => f.endsWith(".md"))
        .map((f) => `docs/modules/llm/providers/${f}`);
    CONFIGURATION["llm-provider-docs.md"].docs = providerDocs;
}

function splitContent(text: string, limit: number): string[] {
    const chunks: string[] = [];
    let remaining = text;

    while (remaining.length > limit) {
        let splitIdx = remaining.lastIndexOf("\n", limit);
        if (splitIdx === -1) {
            splitIdx = limit;
        }
        chunks.push(remaining.slice(0, splitIdx));
        remaining = remaining.slice(splitIdx);
    }
    if (remaining.length > 0) {
        chunks.push(remaining);
    }
    return chunks;
}

function generateFrontmatter(rule: RuleConfig): string {
    const lines = ["---"];
    if (rule.trigger === "always_on") {
        lines.push("trigger: always_on");
        lines.push("globs: **/*"); // Explicit wildmatch for always_on to ensure coverage
    } else {
        lines.push("trigger: glob");
        if (rule.globs && rule.globs.length > 0) {
            const globString = rule.globs.map((g) => `**/${g}`).join(", ");
            lines.push(`globs: ${globString}`);
        } else {
            // Fallback if no globs provided
            lines.push("globs: **/*");
        }
    }
    lines.push("---");
    lines.push("");
    return lines.join("\n");
}

function main() {
    console.log("🚀 Updating agent rules...");

    for (const [ruleFile, config] of Object.entries(CONFIGURATION)) {
        console.log(`Processing ${ruleFile}...`);
        let fullContent = "";

        for (const docFile of config.docs) {
            const docPath = join(PROJECT_ROOT, docFile);
            if (existsSync(docPath)) {
                let content = readFileSync(docPath, "utf8");
                // Fix relative links for rules context
                if (docFile.includes("ARCHITECTURE.md") || ruleFile === "architecture-docs.md") {
                    content = content.replace(/\]\(docs\//g, "](../../docs/");
                }
                fullContent += `\n# Content from ${docFile}\n\n${content}\n`;
            } else {
                console.warn(`⚠️  Warning: Doc file ${docFile} not found.`);
            }
        }

        if (!fullContent.trim()) {
            console.log(`Skipping ${ruleFile} (no content).`);
            continue;
        }

        const frontmatter = generateFrontmatter(config);
        const chunkSize = MAX_CHARS - frontmatter.length - 100;
        const chunks = splitContent(fullContent, chunkSize);

        const baseName = basename(ruleFile, extname(ruleFile));
        const ext = extname(ruleFile);

        chunks.forEach((chunk, index) => {
            const filename = chunks.length > 1 ? `${baseName}-part${index + 1}${ext}` : ruleFile;
            const outPath = join(RULES_DIR, filename);
            writeFileSync(outPath, frontmatter + chunk);
            console.log(`   -> Written ${filename} (${chunk.length} chars)`);
        });
    }

    // Direct copies for critical files (as backup/standard)
    const copyList = [
        { src: "AGENTS.md", dest: "AGENTS.md" },
        { src: "docs/ARCHITECTURE.md", dest: "ARCHITECTURE.md" },
    ];

    copyList.forEach(({ src, dest }) => {
        const srcPath = join(PROJECT_ROOT, src);
        const destPath = join(RULES_DIR, dest);
        if (existsSync(srcPath)) {
            // We still wrap them in the frontmatter so the agent picks them up correctly
            // using the same logic as the loop above (via "agents-md.md" and "architecture-docs.md")
            // but if we want raw copies for direct file access by tools:
            writeFileSync(destPath, readFileSync(srcPath, "utf8"));
            console.log(`   -> Copied ${src} to rules/${dest}`);
        }
    });

    console.log("✅ Rules updated successfully.");
}

main();